### PR TITLE
feat: remove KGH share and modify KGH delegation func

### PR DESCRIPTION
**Pull request by 0xHansLee**
Handling of boosted reward and 1 step undelegation will be covered in the following PR.

Will be merged after merging #347 and #348.

### CODE REVIEW

File: packages/contracts/contracts/L1/AssetManager.sol [hunk 1]
```diff
@@ -169,10 +171,7 @@ contract AssetManager is ISemver, IERC721Receiver, IAssetManager {
         address delegator,
         uint256 tokenId
     ) external view returns (uint128, uint128) {
```
Comment (seolaoh) [inline packages/contracts/contracts/L1/AssetManager.sol:173]:
```suggestion
    ) external view returns (uint128) {
```

File: packages/contracts/contracts/L1/AssetManager.sol [hunk 2]
```diff
@@ -983,36 +765,29 @@ contract AssetManager is ISemver, IERC721Receiver, IAssetManager {
      * @param tokenId   Token Id of the KGH.
      * @param kroInKgh  The amount of KRO in the KGH.
      * @param kroShares The amount of KRO shares to receive.
-     * @param kghShares The amount of KGH shares to receive.
      */
     function _delegateKgh(
```
Comment (seolaoh) [inline packages/contracts/contracts/L1/AssetManager.sol:769]: Please update the comment.

File: packages/contracts/contracts/L1/AssetManager.sol [hunk 3]
```diff
@@ -983,36 +765,29 @@ contract AssetManager is ISemver, IERC721Receiver, IAssetManager {
...
+        KghDelegator storage delegator,
+        uint256 tokenId
+    ) internal returns (uint128, uint128) {
+        uint128 kroInKgh = KGH_MANAGER.totalKroInKgh(tokenId);
+        uint128 kroShares = _convertToKroShares(validator, kroInKgh);
 
         unchecked {
             asset.totalKro += kroInKgh;
             asset.totalKroInKgh += kroInKgh;
             asset.totalKgh += 1;
```
Comment (seolaoh) [inline packages/contracts/contracts/L1/AssetManager.sol:782]:
```suggestion
```

File: packages/contracts/contracts/L1/AssetManager.sol [hunk 4]
```diff
@@ -983,36 +765,29 @@ contract AssetManager is ISemver, IERC721Receiver, IAssetManager {
...
-                kroShares,
-                kghShares
-            );
-        }
 
-        if (kroInKgh > 0) {
-            VALIDATOR_MANAGER.updateValidatorTree(validator, false);
+            ++delegator.kghNum;
+            delegator.kroShares += kroShares;
+            delegator.delegatedAt[tokenId] = block.timestamp;
```
Comment (seolaoh) [inline packages/contracts/contracts/L1/AssetManager.sol:787]:
```suggestion
            delegator.kroShares[tokenId] = kroShares;
            delegator.delegatedAt[tokenId] = uint128(block.timestamp);
```

File: packages/contracts/contracts/L1/AssetManager.sol [hunk 5]
```diff
@@ -363,14 +312,22 @@ contract AssetManager is ISemver, IERC721Receiver, IAssetManager {
...
-        uint128 kroInKgh = KGH_MANAGER.totalKroInKgh(tokenId);
-        uint128 kroShares = _convertToKroShares(validator, kroInKgh);
-        uint128 kghShares = _convertToKghShares(validator);
-        _delegateKgh(validator, tokenId, kroInKgh, kroShares, kghShares);
+    ) external isRegistered(validator) returns (uint128) {
+        // TODO: claim boosted reward and delegate to KRO pool
 
-        emit KghDelegated(validator, msg.sender, tokenId, kroInKgh, kroShares, kghShares);
-        return (kroShares, kghShares);
+        KGH.safeTransferFrom(msg.sender, address(this), tokenId);
```
Comment (seolaoh) [inline packages/contracts/contracts/L1/AssetManager.sol:318]: I wonder why KGH transfer and `updateValidatorTree` comes out from `_delegateKgh`

File: packages/contracts/contracts/L1/AssetManager.sol [hunk 6]
```diff
@@ -379,37 +336,29 @@ contract AssetManager is ISemver, IERC721Receiver, IAssetManager {
...
-            unchecked {
-                kroInKghs += kroInKgh;
-                kroShares += kroSharesForTokenId;
+        Asset storage asset = _vaults[validator].asset;
+        KghDelegator storage delegator = _vaults[validator].kghDelegators[msg.sender];
 
-                ++i;
-            }
+        for (uint256 i = 0; i < tokenIds.length; ) {
+            (kroInKgh, kroShare) = _delegateKgh(validator, asset, delegator, tokenIds[i]);
```
Comment (seolaoh) [inline packages/contracts/contracts/L1/AssetManager.sol:351]: Do you mean `_delegateKghBatch`? Please check the function if it's right.
```suggestion
            (uint128 kroInKgh, uint128 kroShare) = _delegateKgh(validator, asset, delegator, tokenIds[i]);
```

File: packages/contracts/contracts/L1/AssetManager.sol [hunk 7]
```diff
@@ -1022,24 +797,26 @@ contract AssetManager is ISemver, IERC721Receiver, IAssetManager {
      * @param kghCount  The number of KGHs to delegate.
      * @param kroInKghs The amount of KRO in the KGHs.
      * @param kroShares The amount of KRO shares to receive.
-     * @param kghShares The amount of KGH shares to receive.
      */
     function _delegateKghBatch(
```
Comment (seolaoh) [inline packages/contracts/contracts/L1/AssetManager.sol:801]: Please update the comment

Comment (seolaoh) [reply packages/contracts/contracts/L1/AssetManager.sol:801]: I wonder why didn't you unify the input and action between this function and `_delegateKgh`.

File: packages/contracts/contracts/L1/AssetManager.sol [hunk 8]
```diff
@@ -379,37 +336,29 @@ contract AssetManager is ISemver, IERC721Receiver, IAssetManager {
     function delegateKghBatch(
         address validator,
         uint256[] calldata tokenIds
-    ) external isRegistered(validator) returns (uint128, uint128) {
+    ) external isRegistered(validator) returns (uint128) {
         if (tokenIds.length == 0) revert NotAllowedZeroInput();
 
+        // TODO: claim boosted reward and delegate to KRO pool
+
```
Comment (seolaoh) [inline packages/contracts/contracts/L1/AssetManager.sol:343]: KGH transfer is missing

File: packages/contracts/contracts/L1/AssetManager.sol (hunk 6 above)
Comment (seolaoh) [reply packages/contracts/contracts/L1/AssetManager.sol:351]: If you meant to use `_delegateKgh` here also, I think it would be inefficient in the respect of gas. We don't need to update storage every time when using `_delegateKghBatch`.

File: packages/contracts/contracts/L1/AssetManager.sol [hunk 9]
```diff
@@ -430,57 +379,35 @@ contract AssetManager is ISemver, IERC721Receiver, IAssetManager {
     /**
      * @inheritdoc IAssetManager
      */
-    function initUndelegateKgh(address validator, uint256 tokenId) external {
+    function undelegateKgh(address validator, uint256 tokenId) external {
         uint128 kroShares = _vaults[validator].kghDelegators[msg.sender].shares[tokenId].kro;
```
Comment (seolaoh) [inline packages/contracts/contracts/L1/AssetManager.sol:383]: Is it true that you just removed kgh share related things in undelegation functions, not corrected the functions to act correctly?

Comment (seolaoh) [reply packages/contracts/contracts/L1/AssetManager.sol:383]: + I think it's better to remove and add new logics in the same PR in the future.

File: packages/contracts/contracts/L1/AssetManager.sol (hunk 5 above)
Comment (0xHansLee) [reply packages/contracts/contracts/L1/AssetManager.sol:318]: I wanted the `_delegateKgh` function to do the things only related to delegateKgh.
Transferring KGH and updating validator tree are not a roles of `_delegateKgh`.

File: packages/contracts/contracts/L1/AssetManager.sol (hunk 6 above)
Comment (0xHansLee) [reply packages/contracts/contracts/L1/AssetManager.sol:351]: Yes I meant, but you're right. It is gas inefficient when many KGHs are delegated at once. I'll rollback to use `_delegetKghBatch`

File: packages/contracts/contracts/L1/AssetManager.sol (hunk 9 above)
Comment (0xHansLee) [reply packages/contracts/contracts/L1/AssetManager.sol:383]: I see.

File: packages/contracts/contracts/L1/AssetManager.sol [hunk 10]
```diff
@@ -1138,50 +894,25 @@ contract AssetManager is ISemver, IERC721Receiver, IAssetManager {
...
         address validator,
         uint256[] calldata tokenIds,
-        uint128 kroInKghs,
-        uint128 kroShares,
-        uint128 kghShares,
-        uint128 kghAssetsToWithdraw
+        uint128 kroInKghs
     ) internal {
         Vault storage vault = _vaults[validator];
-        uint128 kroAssetsToWithdraw = _convertToKroAssets(validator, kroShares);
```
Comment (seolaoh) [inline packages/contracts/contracts/L1/AssetManager.sol:1153]: Why did you remove it?

File: packages/contracts/contracts/L1/AssetManager.sol [hunk 11]
```diff
@@ -1138,50 +894,25 @@ contract AssetManager is ISemver, IERC721Receiver, IAssetManager {
...
         uint128 boostedRewardsToReceive;
```
Comment (seolaoh) [inline packages/contracts/contracts/L1/AssetManager.sol:904]:
```suggestion
```

File: packages/contracts/contracts/L1/AssetManager.sol [hunk 12]
```diff
@@ -1138,50 +894,25 @@ contract AssetManager is ISemver, IERC721Receiver, IAssetManager {
...
 
         unchecked {
-            vault.asset.totalKroShares -= kroShares;
```
Comment (seolaoh) [inline packages/contracts/contracts/L1/AssetManager.sol:1157]: Why did you remove it?

File: packages/contracts/contracts/L1/AssetManager.sol [hunk 13]
```diff
@@ -1138,50 +894,25 @@ contract AssetManager is ISemver, IERC721Receiver, IAssetManager {
...
-            vault.asset.totalKghShares -= kghShares;
-
             uint128 baseRewardsToReceive = kroAssetsToWithdraw - kroInKghs;
```
Comment (seolaoh) [inline packages/contracts/contracts/L1/AssetManager.sol:907]: reward claim will be handled in claimKghReward, isnt it?
```suggestion
```

**Comment by 0xHansLee** [comment]
@seolaoh I made changes regarding your review in [b46b6a5](https://github.com/kroma-network/kroma/pull/349/commits/b46b6a5720d5d7392b70d846aee2e9de7c7924c0).
In the commit, all codes related to undelegation are rollbacked. I will handle it in the following PR.

File: packages/contracts/contracts/L1/AssetManager.sol [hunk 14]
```diff
@@ -983,35 +905,25 @@ contract AssetManager is ISemver, IERC721Receiver, IAssetManager {
...
-            vault.kghDelegators[msg.sender].shares[tokenId] = KghDelegatorShares(
-                kroShares,
-                kghShares
-            );
-        }
 
-        if (kroInKgh > 0) {
-            VALIDATOR_MANAGER.updateValidatorTree(validator, false);
+            ++delegator.kghNum;
+            delegator.kroShares[tokenId] += kroShares;
```
Comment (seolaoh) [inline packages/contracts/contracts/L1/AssetManager.sol:925]:
```suggestion
            delegator.kroShares[tokenId] = kroShares;
```

File: packages/contracts/contracts/L1/AssetManager.sol [hunk 15]
```diff
@@ -392,10 +350,8 @@ contract AssetManager is ISemver, IERC721Receiver, IAssetManager {
             uint128 kroInKgh = KGH_MANAGER.totalKroInKgh(tokenIds[i]);
             uint128 kroSharesForTokenId = _convertToKroShares(validator, kroInKgh);
 
-            _vaults[validator].kghDelegators[msg.sender].shares[tokenIds[i]] = KghDelegatorShares({
-                kro: kroSharesForTokenId,
-                kgh: kghShares
-            });
+            KghDelegator storage kghDelegator = _vaults[validator].kghDelegators[msg.sender];
```
Comment (seolaoh) [inline packages/contracts/contracts/L1/AssetManager.sol:353]: It's better to move this out of for loop.

File: packages/contracts/contracts/L1/AssetManager.sol [hunk 16]
```diff
@@ -392,10 +350,8 @@ contract AssetManager is ISemver, IERC721Receiver, IAssetManager {
...
+            kghDelegator.delegatedAt[tokenIds[i]] = uint128(block.timestamp);
```
Comment (seolaoh) [inline packages/contracts/contracts/L1/AssetManager.sol:354]:
```suggestion
            kghDelegator.kroShares[tokenId] = kroSharesForTokenId;
            kghDelegator.delegatedAt[tokenIds[i]] = uint128(block.timestamp);
```

File: packages/contracts/contracts/L1/AssetManager.sol [hunk 17]
```diff
@@ -405,11 +361,10 @@ contract AssetManager is ISemver, IERC721Receiver, IAssetManager {
             }
         }
 
-        kghShares *= uint128(tokenIds.length);
-        _delegateKghBatch(validator, uint128(tokenIds.length), kroInKghs, kroShares, kghShares);
+        _delegateKghBatch(validator, uint128(tokenIds.length), kroInKghs, kroShares);
```
Comment (seolaoh) [inline packages/contracts/contracts/L1/AssetManager.sol:365]: Please move update validator tree out of `_delegateKghBatch`

File: packages/contracts/contracts/L1/interfaces/IAssetManager.sol [hunk 18]
```diff
@@ -125,14 +125,12 @@ interface IAssetManager {
...
      * @param tokenIds  Array of token ids of the KGHs.
      * @param kroInKghs The amount of KRO in the KGHs.
-     * @param kroShares The amount of KRO shares received.
      */
     event KghBatchDelegated(
         address indexed validator,
         address indexed delegator,
         uint256[] tokenIds,
-        uint128 kroInKghs,
-        uint128 kroShares
```
Comment (seolaoh) [inline packages/contracts/contracts/L1/interfaces/IAssetManager.sol:135]: Why did you remove it?

File: packages/contracts/contracts/L1/AssetManager.sol [hunk 19]
```diff
@@ -405,11 +361,10 @@ contract AssetManager is ISemver, IERC721Receiver, IAssetManager {
...
 
-        emit KghBatchDelegated(validator, msg.sender, tokenIds, kroShares, kghShares);
-        return (kroShares, kghShares);
+        emit KghBatchDelegated(validator, msg.sender, tokenIds, kroShares);
```
Comment (seolaoh) [inline packages/contracts/contracts/L1/AssetManager.sol:366]:
```suggestion
        emit KghBatchDelegated(validator, msg.sender, tokenIds, kroInKghs);
```

File: packages/contracts/contracts/L1/AssetManager.sol [hunk 20]
```diff
@@ -405,11 +361,10 @@ contract AssetManager is ISemver, IERC721Receiver, IAssetManager {
...
+        return (kroShares);
```
Comment (seolaoh) [inline packages/contracts/contracts/L1/AssetManager.sol:367]:
```suggestion
        return kroShares;
```

File: packages/contracts/contracts/L1/AssetManager.sol (hunk 15 above)
Comment (sm-stack) [inline packages/contracts/contracts/L1/AssetManager.sol:353]: How about declaring this before the for loop? I think it will reduce the gas slightly

File: packages/contracts/contracts/L1/AssetManager.sol [hunk 21]
```diff
@@ -983,35 +905,25 @@ contract AssetManager is ISemver, IERC721Receiver, IAssetManager {
...
-        uint128 kroShares,
-        uint128 kghShares
+        uint128 kroShares
     ) internal {
-        Vault storage vault = _vaults[validator];
-        Asset storage asset = vault.asset;
-
-        KGH.safeTransferFrom(msg.sender, address(this), tokenId);
+        Asset storage asset = _vaults[validator].asset;
+        KghDelegator storage kghDelegator = _vaults[validator].kghDelegators[msg.sender];
```
Comment (seolaoh) [inline packages/contracts/contracts/L1/AssetManager.sol:916]: How about getting `address delegator` as an input instead of `msg.sender`? Also in `_delegateKghBatch`

File: packages/contracts/contracts/L1/AssetManager.sol [hunk 22]
```diff
@@ -1022,24 +934,25 @@ contract AssetManager is ISemver, IERC721Receiver, IAssetManager {
...
-        uint128 kghShares
+        uint128 kroShares
     ) internal {
         Asset storage asset = _vaults[validator].asset;
+        KghDelegator storage kghDelegator = _vaults[validator].kghDelegators[msg.sender];
 
         unchecked {
             asset.totalKro += kroInKghs;
             asset.totalKroInKgh += kroInKghs;
             asset.totalKgh += kghCount;
```
Comment (seolaoh) [inline packages/contracts/contracts/L1/AssetManager.sol:951]:
```suggestion
```

File: packages/contracts/contracts/L1/AssetManager.sol [hunk 23]
```diff
@@ -1022,24 +934,25 @@ contract AssetManager is ISemver, IERC721Receiver, IAssetManager {
...
 
             asset.totalKroShares += kroShares;
-            asset.totalKghShares += kghShares;
+
+            kghDelegator.kghNum += kghCount;
+            kghDelegator.kroShares += kroShares;
```
Comment (seolaoh) [inline packages/contracts/contracts/L1/AssetManager.sol:955]:
```suggestion
```

File: packages/contracts/contracts/L1/AssetManager.sol (hunk 20 above)
Comment (sm-stack) [inline packages/contracts/contracts/L1/AssetManager.sol:367]:
```suggestion
        return kroShares;
```

File: packages/contracts/contracts/L1/AssetManager.sol (hunk 15 above)
Comment (sm-stack) [reply packages/contracts/contracts/L1/AssetManager.sol:353]: Just found that above comments of @seolaoh cover these, I'll resolve them.

File: packages/contracts/contracts/L1/AssetManager.sol (hunk 21 above)
Comment (0xHansLee) [reply packages/contracts/contracts/L1/AssetManager.sol:916]: It's better. I'll make change as you suggested.

File: packages/contracts/contracts/L1/interfaces/IAssetManager.sol (hunk 18 above)
Comment (0xHansLee) [reply packages/contracts/contracts/L1/interfaces/IAssetManager.sol:135]: I mistook it for kghShares. rollback it

**Review by seolaoh — APPROVED** [review]
LGTM

**Review by sm-stack — APPROVED** [review]
